Replace swap_or_assign with move in edm::Wrapper<T> constructor

makortel opened the pull request:
This PR suggests to replace `swap_or_assign()` with move assignment in the constructor of `edm::Wrapper<T>` in order to benefit from compiler-generated move constructor+assignment operator for move-only products instead of having to explicitly write the `swap()` method every time (I got annoyed and after consultation with @Dr15Jones proceeded for a PR). Here the `swap()` was anyway used for a "move" before C++11.

I had to add the move constructor and assignment operators to handful of classes because they were not generated automatically by the compiler (mostly because of deleted copy constructor+assignment operator). There were a few more complex cases:
* `edmNew::DetSetVectorTrans` needed explicit move constructor+assignment operator because of atomic member variable (which are not movable), logic is as in `swap()`
* `MeasurementTrackerEvent` needed explicit move constructor+assignment operator because of the use of a boolean member variable to denote ownership of pointers
* ~~In~~ `LHEEventProduct` ~~I replaced `auto_ptr<PDF>` member with `unique_ptr<PTR>`, after which the compiler-generated move constructor+assignment operator work out of the box~~ needed explicit move constructor+assignment operator because of `auto_ptr`
   * ~~and anyway `auto_ptr` was deprecated in C++11 and removed in C++17~~

For all touched product classes I removed the `swap()` as unnecessary, except for `edmNew::DetSetVector(Trans)` where the `swap()` is used also elsewhere (and `swap()` is still part of the `std::vector` interface so for interface compatibility it makes sense to keep it).


Tested with limited matrix in 10_0_0, no changes expected.

makortel committed:
Replace use of swap_or_assign with std::move

makortel committed:
Add custom move constructor and assignment operator to edmNew::dstvdetails::DetSetVectorTrans!

Needed because std::atomic does not have move constructor (or assignment operator).

makortel committed:
Replace auto_ptr with unique_ptr in LHEEventProduct

makortel committed:
Add explicit move constructor and assignment operator to LHEEventProduct

makortel committed:
Add move constructors and assignment operators to various sim associators

makortel committed:
Add explicit move constructors and assignment operators to LayerHitMapCache, RegionsSeedingHitSets, IntermediateHit{Doublets,Triplets}

makortel committed:
Move move assignment operator to public and add move constructor in JetCorrector

makortel committed:
Enable move constructor and assignment operator in MeasurementTrackerEvent

makortel committed:
Remove swap_or_assign and infrastructure behind it as obsolete

makortel committed:
Remove swap from JetCorrector as unnecessary

makortel committed:
Remove swap from RegionsSeedingHitSets, IntermediateHit{Doublets,Triplets} as unnecessary

makortel committed:
Remove swap from MeasurementTrackerEvent as unnecessary

makortel committed:
Remove swap from sim associators as unnecessary

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22069/3189

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22069/3189/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22069/3189/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

makortel committed:
Fix

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22069/3191

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 6f9256812e14d18a0b60cc58fdc8e9213f1266cc

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22069/25833/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests  RelVals**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test TestGeneratorInterfaceLHEInterfaceFileReading had ERRORS
---> test runtestPhysicsToolsNanoAOD had ERRORS
---> test testTauEmbeddingProducers had ERRORS


* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
1325.7 step2
<pre>runTheMatrix-results/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017/step2_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

makortel committed:
Revert "Replace auto_ptr with unique_ptr in LHEEventProduct"

This reverts commit 309cbcfe76a0c6ba0088d5d870f2e5cc349b62ee.

makortel committed:
Implement explicit move constructor+assignment operator because of auto_ptr member

cmsbuild commented:
The code-checks are being triggered in jenkins.

makortel commented:
Reverted changing `LHEEventProduct::pdf_` from `auto_ptr` to `unique_ptr` (left as later exercise) and implemented move constructor+assignment operator to move the `auto_ptr`.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22069/3200

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #22069 was updated. @perrotta, @smuzaffar, @efeyazgan, @Dr15Jones, @perrozzi, @civanch, @mdhildreth, @monttj, @cmsbuild, @slava77 can you please check and sign again.

cmsbuild commented:
+1
Tested at: ef15111f325b7a63ff35df45d14a54ec756a238d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22069/25845/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22069/25845/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-22069/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 201 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2465763
 - DQMHistoTests: Total failures: 363
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2465231
 - DQMHistoTests: Total skipped: 169
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.04000000005 KiB( 22 files compared)
 - Checked 110 log files, 9 edm output root files, 26 DQM output files

Dr15Jones commented on RecoTracker/MeasurementDet/interface/MeasurementTrackerEvent.h:
```diff
@@ -41,11 +41,8 @@ class MeasurementTrackerEvent {
-
-   //  for the edm wrapper...
-   void swap(MeasurementTrackerEvent &other) ;
+   MeasurementTrackerEvent(MeasurementTrackerEvent && other);
```
This does not work. `MeasurementTrackerEvent` may own the pointers it holds. In that case, the default move constructor will NOT set the pointers to nullptr in the moved object. This will result in the pointed to objects being multiply deleted.

makortel replied:
It actually works because the `~MeasurementTrackerEvent()` deletes the pointers only if `theOwner == true`
https://github.com/cms-sw/cmssw/blob/46354df3148f7d63c56909a3326f57af0343ce25/RecoTracker/MeasurementDet/src/MeasurementTrackerEvent.cc#L4-L9
which is correctly set to `false` for `other`.

But if you want, I can set the pointers of `other` to `nullptr` as well as a precaution.

Dr15Jones replied:
No it is fine. I withdraw my objection.

Dr15Jones commented on RecoTracker/MeasurementDet/src/MeasurementTrackerEvent.cc:
```diff
@@ -10,19 +10,26 @@ MeasurementTrackerEvent::~MeasurementTrackerEvent() {
-        swap(thePixelClustersToSkip, other.thePixelClustersToSkip);
-    }
+MeasurementTrackerEvent::MeasurementTrackerEvent(MeasurementTrackerEvent && other) {
+  theTracker = std::move(other.theTracker);
```
Move does not rest the moved pointer to nullptr.

Dr15Jones commented on SimDataFormats/Associations/interface/MuonToTrackingParticleAssociator.h:
```diff
@@ -33,15 +35,11 @@ namespace reco {
     MuonToTrackingParticleAssociator& operator=( const MuonToTrackingParticleAssociator&) = delete;
 
-    MuonToTrackingParticleAssociatorBaseImpl const* impl_;
+    std::unique_ptr<MuonToTrackingParticleAssociatorBaseImpl const> impl_;
```
Are these stored objects? If so, you have changed the schema which will stop us from being able to read old ROOT files.

makortel replied:
No, they are algorithms stored as transient.
https://github.com/cms-sw/cmssw/blob/38a95c4d8839d94350b8a27837f7fac947e1909e/SimDataFormats/Associations/src/classes_def.xml#L2-L7

Dr15Jones commented on SimDataFormats/Associations/interface/TrackToGenParticleAssociator.h:
```diff
@@ -63,18 +65,14 @@ namespace reco{
     const TrackToGenParticleAssociator& operator=(const TrackToGenParticleAssociator&) = delete; // stop default
     
-    TrackToGenParticleAssociatorBaseImpl* m_impl;
+    std::unique_ptr<TrackToGenParticleAssociatorBaseImpl> m_impl;
```
Schema change

makortel replied:
Transient product.

Dr15Jones commented on SimDataFormats/Associations/interface/TrackToTrackingParticleAssociator.h:
```diff
@@ -104,17 +106,13 @@ namespace reco {
     
     // ---------- member data --------------------------------
-    TrackToTrackingParticleAssociatorBaseImpl* m_impl;
+    std::unique_ptr<TrackToTrackingParticleAssociatorBaseImpl> m_impl;
```
Schema change

makortel replied:
This is transient as well.

Dr15Jones commented on SimDataFormats/Associations/interface/VertexToTrackingVertexAssociator.h:
```diff
@@ -27,17 +29,13 @@ namespace reco {
     
     // ---------- member data --------------------------------
-    VertexToTrackingVertexAssociatorBaseImpl* m_impl;
+    std::unique_ptr<VertexToTrackingVertexAssociatorBaseImpl> m_impl;
```
Schema change

makortel replied:
This is transient as well
https://github.com/cms-sw/cmssw/blob/38a95c4d8839d94350b8a27837f7fac947e1909e/SimDataFormats/Associations/src/classes_def.xml#L9-L10

Dr15Jones commented:
+1

perrotta commented:
@makortel : jenkins tests show a lot of differences in several muon related objects in the TTbar PU wf 25202 (it seems that the difference just roots on RPCs, and then propagates to all higher level objects). See https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_1_X_2018-01-31-2300+22069/24871/validateJR/all_OldVSNew_TTbarPUwf25202p0/

Looking at the files you touched in this PR, the only one related to muons seems to be the MuonToTrackingParticleAssociator: but this seems to me only linked to the validation, rather than to the simulation or reco steps.

Maybe the differences come from the MB event samples used for the PU, although it is not obvious to me why it should only show up in muons, and not in Tracker, etc

Could you please have a look?

makortel commented:
@perrotta Thanks for report, I'll take a look (on the top of my head I have no idea).

I went through all alternative-comparisons, and in addition to 25202.0 also 136.788 showed differences in RPC RecHits (but not elsewhere).

slava77 commented:
On 2/8/18 12:54 AM, perrotta wrote:
> @makortel <https://github.com/makortel> : jenkins tests show a lot of
> differences in several muon related objects in the TTbar PU wf 25202 (it
> seems that the difference just roots on RPCs, and then propagates to all
> higher level objects). See
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_1_X_2018-01-31-2300+22069/24871/validateJR/all_OldVSNew_TTbarPUwf25202p0/
> 
> Looking at the files you touched in this PR, the only one related to
> muons seems to be the MuonToTrackingParticleAssociator: but this seems
> to me only linked to the validation, rather than to the simulation or
> reco steps.
> 
> Maybe the differences come from the MB event samples used for the PU,
> although it is not obvious to me why it should only show up in muons,
> and not in Tracker, etc
> 
> Could you please have a look?
> 

Is this at least reproducible?

makortel commented:
> Is this at least reproducible?

I'm going to test if it is, but I'm not quite there yet.

slava77 commented:
> I went through all alternative-comparisons, and in addition to 25202.0 also 136.788 showed differences in RPC RecHits (but not elsewhere).

there are no changes in 136.788  in reco. I'd blame the DQM plot differences in this case on randomness of RPC DQM which I've seen rather frequently in the past.

makortel commented:
@cmsbuild, please test

I was not able to reproduce in CMSSW_10_1_X_2018-02-07-2300, so let's try again here as well.

cmsbuild commented:
The tests are being triggered in jenkins.

makortel commented:
@cmsbuild, please abort

Looks to me that running the tests failed for some reason (*), restarting.

(*) from "Details"
```
23:09:49 FATAL: command execution failed
23:09:49 java.io.EOFException
...
23:09:49 Build step 'Execute shell' marked build as failure
...
23:09:49 ERROR: Step ‘E-mail Notification’ failed: no workspace for ib-any-integration #25984
23:09:49 ERROR: Step ‘Delete workspace when build is done’ failed: no workspace for ib-any-integration #25984
23:09:50 Finished: FAILURE
```

cmsbuild commented:
Jenkins tests are aborted.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

makortel commented:
Didn't find anything relevant with valgrind from step3 of 25202.0.

cmsbuild commented:
-1

Tested at: ef15111f325b7a63ff35df45d14a54ec756a238d

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22069/25991/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test TestIOMCRandomEngineService had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22069/25991/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-22069/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017+HARVESTNANOAODMC2017

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 201 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2464162
 - DQMHistoTests: Total failures: 361
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2463632
 - DQMHistoTests: Total skipped: 169
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.04999999981 KiB( 22 files compared)
 - Checked 111 log files, 9 edm output root files, 27 DQM output files

makortel commented:
The unit test fails with
```
RandomNumberGeneratorService multistream test replay from text file
==============================================
Error: Text file containing states not found
status = 256
```
I have no idea how that could be caused by this PR (seems to be ok in the IB). Temporary glitch?

makortel commented:
The RPC differences appear to be reproducible with the PR tests at least.

makortel commented:
Let me ask for a clarification in the "comparison vs. baseline". Is the "baseline" picked directly from the IB, i.e. in case of 25202.0 it takes the GEN-SIM from existing 9_2_4 dataset instead of re-running step1 (as is done for IB+PR)?

(mostly for @smuzaffar, I guess)

smuzaffar commented:
@makortel , baseline is picked up from IB directly. For most of workflows (including 25202.0) we re-run step1 and do not re-use existing data.

davidlange6 commented:
you can navigate to them from the summary page - eg, eventually you get to this 

https://cmssdt.cern.ch/SDT/jenkins-artifacts/ib-baseline-tests/CMSSW_10_1_X_2018-02-08-2300/slc6_amd64_gcc630/-GenuineIntel/matrix-results/

> On Feb 12, 2018, at 10:53 AM, Matti Kortelainen <notifications@github.com> wrote:
> 
> Let me ask for a clarification in the "comparison vs. baseline". Is the "baseline" picked directly from the IB, i.e. in case of 25202.0 it takes the GEN-SIM from existing 9_2_4 dataset instead of re-running step1 (as is done for IB+PR)?
> 
> (mostly for @smuzaffar, I guess)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

makortel commented:
Thanks @smuzaffar and @davidlange6. David's link was more or less what I was trying to look for but failed to notice so far. So indeed step1 is rerun also for baseline. (I'm just trying to think how the PR test differs from my private test)

makortel commented:
After playing around a bit I'm coming back to @perrotta's observation in https://github.com/cms-sw/cmssw/pull/22069#issuecomment-364043741 that the MinBias events for pileup are different between IB and IB+PR (IB starts with `765D30CA-7C62-E711-A735-0025905B856E.root`, IB+PR uses `D4CE40F8-7862-E711-85D5-003048FFD772.root` for the first event and then continues with `765D30CA-7C62-E711-A735-0025905B856E.root`). But why this difference in step2 affects only RPC I have no idea.

I took a look on a random no-changes PR tested recently (#22164), and also there the MinBias files for pileup are different for the first event between IB (*) and IB+PR (**). Curiously the DQM histograms are nevertheless the same. 

(*) https://cmssdt.cern.ch/SDT/jenkins-artifacts/ib-baseline-tests/CMSSW_10_1_X_2018-02-09-1100/slc6_amd64_gcc630/-GenuineIntel/matrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/step2_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25.log
(**) https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22164/25995/runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/step2_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25.log

In fact I can see similar behaviour in 10224.0 in this PR as well, the MinBias file for first event is different between IB and IB+PR, and for the remaining events the files are the same, and no difference in DQM histograms.

So what makes 25202.0 of this PR different from the other two cases is that the MinBias file that IB+PR uses from 2nd event onwards gets used by IB starting from 1st event.

I'm very puzzled by this behaviour of MinBias pileup files (both in general and what happens with this PR), so I'm likely missing something. But my feeling is that the reason for the observed differences with RPC lies somewhere there.

makortel commented:
@cmsbuild, please test

Let's see again what happens.

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
Mixing is a bit of a strange beast. We need to know what data products will be available from the mixing files before we know what random number to use. So we open a file randomly, look at what data products are available. That sets what data products will be allowed to be read from all the mixing files. We then close that file. Then when we can get the actual random generator to use during begin Lumi we use that generator to pick the actual file and event to start to use for processing. 

Therefore the first open file basically has no effect on the results since it is only used to read the names of the branches in the file.

makortel commented:
@Dr15Jones Ok, thanks. I guess I got a bit confused by the printouts
```
Opened file
Begin processing 1st record
Closed file
Opened another file
Begin processing 2nd record
```
that the 1st file is closed and 2nd opened before the 1st event is actually processed.

So the mystery continues...

slava77 commented:
Is it now clear that the step2.root files are different?

- if the baseline and this PR jenkins step2.root are used to run step3/4
with the same baseline code, is there a difference in the RPC outputs in
reco?



On 2/14/18 2:06 AM, Matti Kortelainen wrote:
> After playing around a bit I'm coming back to @perrotta
> <https://github.com/perrotta>'s observation in #22069 (comment)
> <https://github.com/cms-sw/cmssw/pull/22069#issuecomment-364043741> that
> the MinBias events for pileup are different between IB and IB+PR (IB
> starts with |765D30CA-7C62-E711-A735-0025905B856E.root|, IB+PR uses
> |D4CE40F8-7862-E711-85D5-003048FFD772.root| for the first event and then
> continues with |765D30CA-7C62-E711-A735-0025905B856E.root|). But why
> this difference in step2 affects only RPC I have no idea.
> 
> I took a look on a random no-changes PR tested recently (#22164
> <https://github.com/cms-sw/cmssw/pull/22164>), and also there the
> MinBias files for pileup are different for the first event between IB
> (*) and IB+PR (**). Curiously the DQM histograms are nevertheless the same.
> 
> (*)
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/ib-baseline-tests/CMSSW_10_1_X_2018-02-09-1100/slc6_amd64_gcc630/-GenuineIntel/matrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/step2_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25.log
> (**)
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22164/25995/runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/step2_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25.log
> 
> In fact I can see similar behaviour in 10224.0 in this PR as well, the
> MinBias file for first event is different between IB and IB+PR, and for
> the remaining events the files are the same, and no difference in DQM
> histograms.
> 
> So what makes 25202.0 of this PR different from the other two cases is
> that the MinBias file that IB+PR uses from 2nd event onwards gets used
> by IB starting from 1st event.
> 
> I'm very puzzled by this behaviour of MinBias pileup files (both in
> general and what happens with this PR), so I'm likely missing something.
> But my feeling is that the reason for the observed differences with RPC
> lies somewhere there.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/22069#issuecomment-365555058>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbojY5KSzZfiq1T5aQvpkPqSmLyePks5tUrAJgaJpZM4R1Yn_>.
>

makortel commented:
@slava77 Earlier I thought I would have had evidence suggesting difference in step2, but taking a more careful look now I can say that the IB vs. IB+PR step2 root files look identical (by running the same step3+4 and diffing the DQM files).

Sorry for the noise...

makortel commented:
Said that, if anyone has good ideas what to check, I'm all ears (as I'm running out of ideas).

slava77 commented:
On 2/14/18 7:01 AM, Matti Kortelainen wrote:
> @slava77 <https://github.com/slava77> Earlier I thought I would have had
> evidence suggesting difference in step2, but taking a more careful look
> now I can say that the IB vs. IB+PR step2 root files look identical (by
> running the same step3+4 and diffing the DQM files).
> 
> Sorry for the noise...
> 


Just to make sure:
take
- A: jenkins baseline step2.root
- B: jenkins PR step2.root

- 1: run locally step3,4 in IB
- 2: run locally step3,4 in IB+PR

Do I understand correctly that the claim is
A1 = A2 = B1 = B2 ?

makortel commented:
@slava77 I've tested that A2 == B2. I'll repeat the test with IB-only and let's see.

cmsbuild commented:
-1

Tested at: ef15111f325b7a63ff35df45d14a54ec756a238d

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22069/26062/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test TestIOMCRandomEngineService had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22069/26062/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-22069/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017+HARVESTNANOAODMC2017

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 201 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2465360
 - DQMHistoTests: Total failures: 361
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2464830
 - DQMHistoTests: Total skipped: 169
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.30000000003 KiB( 22 files compared)
 - Checked 111 log files, 9 edm output root files, 27 DQM output files

makortel commented:
@slava77 Ok, it seems that when running locally indeed "A1 = A2 = B1 = B2"...

The last round of tests repeated the previous ones, i.e. RandomEngineService unit test fails, differences in RPC in 25202.0. I begin to wonder if the two issues could be connected, I'll take a closer look on that tomorrow.

slava77 commented:
On 2/14/18 12:19 PM, Matti Kortelainen wrote:
> @slava77 <https://github.com/slava77> Ok, it seems that when running
> locally indeed "A1 = A2 = B1 = B2"...
> 
> The last round of tests repeated the previous ones, i.e.
> RandomEngineService unit test fails, differences in RPC in 25202.0. I
> begin to wonder if the two issues could be connected, I'll take a closer
> look on that tomorrow.
> 
> —

At this point (or very soon), I'd probably make an issue on github and
move on.

makortel commented:
Further investigation revealed that the RandomService unit test and the RPC issue are not related. The unit test is fixed in #22231.

makortel commented:
I finally managed to reproduce the RPC issue by running `runTheMatrix.py -l 25202.0 --ibeos`. The `--ibeos` was necessary (for some reason) to get the step2 (i.e. pileup mixing) to use exactly the same MinBias files as the PR tests.

perrotta commented:
Thank you Matti for the hint: I did not know the flag "--ibeos" (which corresponds to "Use IB EOS site configuration").

I tested the same using that flag, and I verified that with such a flag identical python configs are generated for step2 and step3 between baseline and baseline+thisPR (although being different from the ones obtained without "--ibeos", the difference being in the list of MB files used for premixing [1]). From the output logs, the very first file used in the pre-mixing is different in the logs of the baseline and the ones obtained in baseline+thisPR, and this seems to be the only difference. One must say, however, that also without setting the flag "--ibeos" the very first file that appears in the logs is different between baseline and baseline+thisPR..

Checking the outputs, with the configurations obtained with the runTheMatrix command with that flag set I can reproduce the different distributions in the RPC between baseline and baseline+thisPR, i.e. (following the definitions in https://github.com/cms-sw/cmssw/pull/22069#issuecomment-365636652)
A1 != B2
Then I verified that
(A1 == A2) != (B1 == B2)
Therefore the difference comes from the step2, while step3 reproduces identically once the step2.root provided in input is the same. 

The conclusions I can draw are therefore the following:
- The difference in RPC distributions comes from step2
- It comes from something which is in the release itself, and not from the python configurations used to produce the step2
- It only shows up if "--ibeos" is set in runTheMatrix
  - (Perhaps it was just triggered by a particular "unfortunate" order of the files which ended up in this test, and it has nothing to do with the content of this PR?)


[1] Even if I compare the edmConfigDumps of the configs, the only difference between using or not using "--ibeos" is in the order of the files included with
```
    input = cms.SecSource("EmbeddedRootSource",
        fileNames = cms.untracked.vstring(...),
```

Dr15Jones commented:
@perrotta The first files opened by the mixing modules are just used to determine what data products (i.e. TBranches) the files have. Those files are then closed and new files chosen by using the RandomNumberService are opened. It is the second group of files which actually have their data read.

Therefore jobs differing just by those initial files should have no effect.

perrotta commented:
Thank you @Dr15Jones, you already expained it well in  
https://github.com/cms-sw/cmssw/pull/22069#issuecomment-365600008

I just reported the observations I made as a results of a few tests,  
namely that the configuration is identical in baseline and  
baseline+PR, with the only difference of that first file, that as you  
explains is only needed to identify teh data products in it. That's  
why I imagine that if there is something causing the different  
behaviour in the RPC must be in the CMSSW code itself, and not in the  
configuration. Unfortunately, this is the only conclusion I can draw  
by now, but I wanted to share anyhow just in case there is something  
in it that can enlighten someone else.

Chris Jones <notifications@github.com> ha scritto:

> @perrotta The first files opened by the mixing modules are just used  
> to determine what data products (i.e. TBranches) the files have.  
> Those files are then closed and new files chosen by using the  
> RandomNumberService are opened. It is the second group of files  
> which actually have their data read.
>
> Therefore jobs differing just by those initial files should have no effect.
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/cms-sw/cmssw/pull/22069#issuecomment-366450821

perrotta commented:
@cmsbuild, please test
(With #22231 merged the unit test failure should have gone)

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: ef15111f325b7a63ff35df45d14a54ec756a238d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22069/26153/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22069/26153/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 205 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2465360
 - DQMHistoTests: Total failures: 361
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2464830
 - DQMHistoTests: Total skipped: 169
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.09999999998 KiB( 22 files compared)
 - Checked 111 log files, 9 edm output root files, 27 DQM output files

makortel commented:
@perrotta I'm already investigating what exactly goes differently for RPC in step2 between IB and IB+PR. So far I can confirm that I see differences, so it's "only" a matter of tracing the origin.

(regarding the unit test, this time the "erroneous" condition of "no events in stream 2" did not occur
```
RandomNumberGeneratorService multistream
==============================================
Begin processing the 1st record. Run 1, Event 1, LumiSection 1 on stream 0 at 19-Feb-2018 00:24:29.978 CET
Begin processing the 2nd record. Run 1, Event 2, LumiSection 1 on stream 1 at 19-Feb-2018 00:24:29.979 CET
Begin processing the 3rd record. Run 1, Event 3, LumiSection 1 on stream 2 at 19-Feb-2018 00:24:29.979 CET
Begin processing the 4th record. Run 1, Event 4, LumiSection 2 on stream 0 at 19-Feb-2018 00:24:30.244 CET
Begin processing the 5th record. Run 1, Event 5, LumiSection 2 on stream 1 at 19-Feb-2018 00:24:30.244 CET
```
)

makortel commented:
I've summarized my findings so far to issue #22266. In short, `MixCollection<T>` that `RPCDigiProducer` and `RPCDigitizer` use to loop over `PSimHit`s in `CrossingFrame<PSimHit>`s has undefined behaviour in its iterator's `MixItr::operator!=()`. Further, in practice it can happen that if the memory of two vectors if `CrossingFrame<T>` are ordered in a given way, `MixCollection<T>::begin() ==  MixCollection<T>::end()`.

I'm trying make a simple patch to make `MixItr::operator!=()` defined, but I'm not sure if I should push it here or make a separate PR (in principle it can cause changes elsewhere).

Since the bug in `MixCollection<T>` depends on memory orderings, I have no idea why exactly it is always 25202.0 in this PR that shows the difference. Probably changing the `Event::put()` to use (compiler-generated) move instead of the implemented `swap()` did something for the memory ordering to be triggered in this PR.

perrotta commented:
Good catch @makortel !

I would suggest making another PR with your patch, since it is a fix not fully related to this one.
But once ready, I would be curious to test the two together to see what is their overall effect

makortel commented:
I have a minimal fix in #22270. I tested 25202.0 (with `--ibeos`) in IB+#22270 and IB+#22270+this PR, and there is no difference anymore from this PR.

perrotta commented:
please test with #22270

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: ef15111f325b7a63ff35df45d14a54ec756a238d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22069/26189/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/6138cf4161c2bb4b2943f9996630dd60101b20b2
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22069/26189/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22069/26189/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22069/26189/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 28
 - DQMHistoTests: Total histograms compared: 2498032
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2497855
 - DQMHistoTests: Total skipped: 176
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.05999999998 KiB( 23 files compared)
 - Checked 115 log files, 9 edm output root files, 28 DQM output files

fabiocos commented:
@civanch @perrozzi @efeyazgan could you please check?

perrozzi commented on SimDataFormats/GeneratorProducts/interface/LHEEventProduct.h:
```diff
@@ -24,6 +24,27 @@ class LHEEventProduct {
 	LHEEventProduct(const lhef::HEPEUP &hepeup,
 			const double originalXWGTUP) : 
 	  hepeup_(hepeup), originalXWGTUP_(originalXWGTUP) {}
+	LHEEventProduct(LHEEventProduct&& other) {
```
may I ask what is addition used for?

makortel replied:
`LHEEventProduct(LHEEventProduct&& other)` is a move constructor and `LHEEventProduct& operator=(LHEEventProduct&& other)` is a move assignment operator. They need to be explicitly defined for `LHEEventProduct` because of `auto_ptr` member.

Motivation (from PR description) is to replace the current pre-C++11 "swap-or-copy" idiom inside `edm::Wrapper` with move in order to not require move-only products to define `swap()` when compiler-generated move constructor and assignment operator are enough.

perrozzi commented:
+1

civanch commented:
+1

fabiocos commented:
+1

fabiocos commented:
merge

SanghyunKo commented on SimDataFormats/GeneratorProducts/interface/LHEEventProduct.h:
```diff
@@ -24,6 +24,27 @@ class LHEEventProduct {
+	  originalXWGTUP_ = std::move(originalXWGTUP_);
+	  scales_ = std::move(scales_);
+	  npLO_ = std::move(npLO_);
+	  npNLO_ = std::move(npNLO_);
```
Hi @makortel , I'm MC generator validator of PdmV group. While validating 10_1_0_pre2, I've found that some informations such as npLO & npNLO are missing in generated MC samples. It seems that a bug is introduced from this PR, especially SimDataFormats package. I tried adding `other`. in L32-35 & L42-45, but simply adding it could not fix the issue. May I ask you to review this PR? Any questions would be appreciated.

makortel replied:
@SanghyunKo Whoops, that's certainly a bug (sorry about that). But I'm puzzled why adding `other.` on the right-hand sides of lines 32-35 and 42-45 would not fix the issue. Could you give me a (simple) recipe how to check myself whether  the bug is fixed or not?

makortel replied:
@SanghyunKo 
I see the `other` was already added by #22429 (merged in 10_1_0_pre3). I'm nevertheless interested to double-check if the fix actually works or not, so 

> Could you give me a (simple) recipe how to check myself whether the bug is fixed or not?

SanghyunKo replied:
@makortel Here I attached link to dropbox that contains simple command that I used to test it, python configuration file and some example of generated LHE file in pre1 & pre2 respectively. I'm also puzzled but let's cross-check it first.

[Link to dropbox](https://www.dropbox.com/s/wvwsmlqw9tvzmha/npLOandnpNLO_v2.zip?dl=0)

makortel replied:
@SanghyunKo 
Thanks for letting me know you got it to work as well (I can also confirm it with the test you provided).
